streams/src/main/java/org/apache/kafka/streams/processor/internals/assignment/DefaultStateConstrainedBalancedAssignor.java

```diff
@@ -0,0 +1,218 @@
+     *
+     * @param assignment assignment
+     * @param unassignedTasksWithoutCaughtUpClients list of task that could not be assigned since no caught-up client exists for them
+     * @param statefulTasksToRankedClients ranked clients map
+     */
+    private void assignTasksWithoutCaughtUpClients(final Map<String, List<TaskId>> assignment,
+                                                   final List<TaskId> unassignedTasksWithoutCaughtUpClients,
+                                                   final Map<TaskId, List<ClientIdAndLag>> statefulTasksToRankedClients) {
+        for (final TaskId taskId : unassignedTasksWithoutCaughtUpClients) {
+            List<ClientIdAndLag> clientIdAndLags = statefulTasksToRankedClients.get(taskId);
+            int previousLag = clientIdAndLags.get(0).lag();
+            String clientWithLeastTasks = null;
+            int taskCount = Integer.MAX_VALUE;
+            int i = 0;
+            while (clientIdAndLags.get(i).lag() == previousLag) {
+                final String clientId = clientIdAndLags.get(i).clientId();
+                final int assignedTasksCount = assignment.get(clientId).size();
+                if (taskCount > assignedTasksCount) {
+                    clientWithLeastTasks = clientId;
+                    taskCount = assignedTasksCount;
+                }
+                ++i;
+            }
+            assignment.get(clientWithLeastTasks).add(taskId);
+        }
+    }
+
+    /**
+     * Balance the assignment.
+     *
+     * @param assignment assignment
+     * @param balanceFactor balance factor
+     * @param statefulTasksToRankedClients ranked clients map
+     * @param tasksToCaughtUpClients map from task IDs to lists of caught-up clients
+     */
+    private void balance(final Map<String, List<TaskId>> assignment,
+                         final int balanceFactor,
+                         final Map<TaskId, List<ClientIdAndLag>> statefulTasksToRankedClients,
+                         final Map<TaskId, List<String>> tasksToCaughtUpClients) {
+        final List<String> clients = new ArrayList<>(assignment.keySet());
+        if (clients.isEmpty()) {
+            throw new IllegalStateException("Clients within an assignment must not be empty");
+        }
+        Collections.sort(clients);
+        for (final String client1 : clients) {
+            for (final String client2 : clients) {
+                final List<TaskId> taskIds1 = assignment.get(client1);
+                final List<TaskId> taskIds2 =  assignment.get(client2);
+                final List<TaskId> source = taskIds1.size() > taskIds2.size() ? taskIds1 : taskIds2;
+                final List<TaskId> destination = source == taskIds1 ? taskIds2 : taskIds1;
+                while (
+                    source.size() - destination.size() > balanceFactor &&
+                    !tasksToCaughtUpClients.containsKey(source.get(source.size() - 1))
```

Comment:
prop: move this condition to a descriptively named and/or commented helper function, eg `hasMoveableTask`
 -- also would be nice to include brief explanation as to why we check the last task only (because all tasks that are caught-up come before the moveable/un-caught-up tasks?)

Reply:
maybe also add a `lastAssignedTask(List<TaskId>)` helper to clean up `source.get(source.size() - 1)` used here and below

Reply:
Ack